src/storage/gc_worker.rs

```diff
@@ -186,113 +178,33 @@ impl<E: Engine> GCRunner<E> {
-                    "region_id" => ctx.get_region_id(),
-                    "versions" => gc_info.found_versions,
-                    "key" => %k
-                );
-            }
-            // TODO: we may delete only part of the versions in a batch, which may not beyond
-            // the logging threshold `GC_LOG_DELETED_VERSION_THRESHOLD`.
-            if gc_info.deleted_versions as usize >= GC_LOG_DELETED_VERSION_THRESHOLD {
-                debug!(
-                    "GC deleted plenty versions for a key";
-                    "region_id" => ctx.get_region_id(),
-                    "versions" => gc_info.deleted_versions,
-                    "key" => %k
-                );
-            }
-
-            if !gc_info.is_completed {
-                next_scan_key = Some(k);
-                break;
-            }
-        }
-        self.stats.add_statistics(&txn.take_statistics());
-
-        let modifies = txn.into_modifies();
-        if !modifies.is_empty() {
-            self.engine.write(ctx, modifies)?;
+        if !reader.need_gc(safe_point, self.ratio_threshold) {
+            debug!("skip gc region {}", ctx.get_region_id());
+            return Ok(());
         }
-        Ok(next_scan_key)
-    }
-
-    fn gc(&mut self, ctx: &mut Context, safe_point: u64) -> Result<()> {
-        debug!(
-            "start doing GC";
-            "region_id" => ctx.get_region_id(),
-            "safe_point" => safe_point
-        );
 
-        let mut next_key = None;
-        loop {
-            // Scans at most `GC_BATCH_SIZE` keys
-            let (keys, next) = self
-                .scan_keys(ctx, safe_point, next_key, GC_BATCH_SIZE)
-                .map_err(|e| {
-                    warn!("gc scan_keys failed"; "region_id" => ctx.get_region_id(), "safe_point" => safe_point, "err" => ?e);
-                    e
-                })?;
-            if keys.is_empty() {
-                break;
-            }
-
-            // Does the GC operation on all scanned keys
-            next_key = self.gc_keys(ctx, safe_point, keys, next).map_err(|e| {
-                warn!("gc gc_keys failed"; "region_id" => ctx.get_region_id(), "safe_point" => safe_point, "err" => ?e);
-                e
-            })?;
-            if next_key.is_none() {
-                break;
-            }
+        let mut raft_cmd = raft_cmdpb::RaftCmdRequest::new();
+        raft_cmd.mut_header().set_region_id(ctx.region_id);
+        raft_cmd
+            .mut_header()
+            .set_region_epoch(ctx.get_region_epoch().clone());
+        raft_cmd.mut_header().set_peer(ctx.get_peer().clone());
+
+        let mut request = raft_cmdpb::AdminRequest::new();
+        request.set_cmd_type(raft_cmdpb::AdminCmdType::MvccGc);
+        request.mut_mvcc_gc().set_safe_point(safe_point);
+        raft_cmd.set_admin_request(request);
+
+        let router = self.raft_store_router.as_ref().unwrap();
+        if let Err(e) = router.send_command(raft_cmd, RaftCallback::None) {
```

Comment:
Emmm... Could we perform GC directly on RocksDB independently on each node, regard less of the Raft layer?

Reply:
I think we can. However it needs more change in this file, so I think we can do it in another PR.

Reply:
Compact by region It may cause write amplification. Could you take one screenshot of IO and rocksdb compaction flow?

Reply:
Here is the doc: https://docs.google.com/document/d/1KQWKGW0AABGZ2_DitPXO-zMuym2kBqz3OjUXbCB68lY/edit?ts=5d5b64e5.
From the test result, we have 2 improvements to reduce write amplification:
- collecting at least 16 regions to performance 1 compaction
- compact-range ignoring memtable and level0